### PR TITLE
util: add linkat to syscall sandbox (AllowFileSystem)

### LINKED ISSUE #24536: Syscall Sandbox Termination
I was testing v23.0rc2 compiled.

`./configure --without-miniupnpc --without-natpmp --enable-hardening --without-bdb --with-qrencode --with-zmq --with-sqlite=yes --disable-external-signer`

bitcoin.conf:

> assumevalid=0
> blockfilterindex=1
> coinstatsindex=1
> dbcache=1000
> prune=0
> sandbox=log-and-abort
> txindex=1
> cjdnsreachable=1
> discover=1
> i2psam=127.0.0.1:7656
> listen=1
> listenonion=1
> networkactive=1
> peerblockfilters=1
> proxy=127.0.0.1:9050
> addresstype=bech32m
> avoidpartialspends=1
> changetype=bech32m
> maxapsfee=0.001
> walletrbf=1
> shrinkdebugfile=1

I deleted my data directory (/home/user/.bitcoin) to start a fresh download of the blockchain to test any errors.  Ran out of space, and after deleting more and trying to start bitcoin-qt again:

> ERROR: The syscall "linkat" (syscall number 265) is not allowed by the syscall sandbox in thread "main". Please report.
> terminate called without an active exception
> Aborted

and in debug.log:

> 2022-03-11T13:00:02Z init message: Verifying blocks…
> 2022-03-11T13:00:02Z Verifying last 6 blocks at level 3
> 2022-03-11T13:00:02Z [0%]...[16%]...[33%]...[50%]...[66%]...[83%]...[99%]...[DONE].
> 2022-03-11T13:00:02Z No coin database inconsistencies in last 6 blocks (2938 transactions)
> 2022-03-11T13:00:02Z  block index            4963ms
> 2022-03-11T13:00:02Z Opening LevelDB in /home/user/.bitcoin/indexes/txindex
> 2022-03-11T13:00:02Z Opened LevelDB successfully
> 2022-03-11T13:00:02Z Using obfuscation key for /home/user/.bitcoin/indexes/txindex: 0000000000000000
> 2022-03-11T13:00:02Z Opening LevelDB in /home/user/.bitcoin/indexes/blockfilter/basic/db
> 2022-03-11T13:00:02Z txindex thread start
> 2022-03-11T13:00:02Z txindex is enabled at height 326934
> 2022-03-11T13:00:02Z txindex thread exit
> 2022-03-11T13:00:02Z Opened LevelDB successfully
> 2022-03-11T13:00:02Z Using obfuscation key for /home/user/.bitcoin/indexes/blockfilter/basic/db: 0000000000000000
> 2022-03-11T13:00:02Z Opening LevelDB in /home/user/.bitcoin/indexes/coinstats/db
> 2022-03-11T13:00:02Z basic block filter index thread start
> 2022-03-11T13:00:02Z basic block filter index is enabled at height 326934
> 2022-03-11T13:00:02Z basic block filter index thread exit
> 2022-03-11T13:00:02Z Opened LevelDB successfully
> 2022-03-11T13:00:02Z Using obfuscation key for /home/user/.bitcoin/indexes/coinstats/db: 0000000000000000
> 2022-03-11T13:00:02Z ERROR: Init: Cannot read current coinstatsindex state; index may be corrupted
> 2022-03-11T13:00:02Z Shutdown: In progress...
> 2022-03-11T13:00:02Z scheduler thread exit
> 2022-03-11T13:00:02Z Shutdown: done
> 2022-03-11T13:00:02Z ERROR: The syscall "linkat" (syscall number 265) is not allowed by the syscall sandbox in thread "main". Please report.

This was done on a Qubes 4.0 machine in a Debian 11 VM.

### PR DESCRIPTION
**Pull request by fanquake**
Should fix #24536.

### CODE REVIEW

**Comment by MarcoFalke** [comment]
cr ACK 9809db3577f0fa618bea42635b1581e628a30395

but might be good to test

**Comment by luke-jr** [comment]
Should we allow regular `link` too? My manpage says glibc can fallback to `link` for emulating `linkat` on older kernels.

**Comment by fanquake** [comment]
> Should we allow regular link too? My manpage says glibc can fallback to link for emulating linkat on older kernels.

Yes, if someone reports an issue with it. I don't think we want/need to preemptively allow syscalls.

**Comment by luke-jr** [comment]
We don't usually leave known bugs until someone reports them...

**Comment by MarcoFalke** [comment]
Yeah, I think it is fine to put in both.

**Comment by fanquake** [comment]
> We don't usually leave known bugs until someone reports them...

Sure, but this isn't a known bug yet. If we want it to be one, can you elaborate more than "my manage says maybe this is needed for some older kernels".

Which manpage? Which older kernel (versions)? Which versions of glibc perform the fallback-ing? Do we support them? Can you reproduce the bug?

I'm not really interested in adding exceptions to our syscall sandbox based on hypotheticals derived from a possibly irrelevant (to our project & runtime back-compatibilities) sentence in a manpage. Especially if the problem hasn't actually been run into in the wild.

File: src/util/syscall_sandbox.cpp [hunk 1]
```diff
@@ -592,6 +592,7 @@ class SeccompPolicyBuilder
         allowed_syscalls.insert(__NR_getcwd);          // get current working directory
         allowed_syscalls.insert(__NR_getdents);        // get directory entries
         allowed_syscalls.insert(__NR_getdents64);      // get directory entries
+        allowed_syscalls.insert(__NR_linkat);          // create relative to a directory file descriptor
```
Comment (hebasto) [inline src/util/syscall_sandbox.cpp:595]: Tested on Ubuntu 22.04:
```suggestion
        allowed_syscalls.insert(__NR_inotify_rm_watch);// remove an existing watch from an inotify instance
        allowed_syscalls.insert(__NR_linkat);          // create relative to a directory file descriptor
```

**Comment by Rspigler** [comment]
Tested ACK - this fixes https://github.com/bitcoin/bitcoin/issues/24536

**Comment by MarcoFalke** [comment]
Going to merge this, since it is tested and reviewed. Maybe issues on other OS can be fixed in a separate commit?

File: src/util/syscall_sandbox.cpp (hunk 1 above)
Comment (hebasto) [reply src/util/syscall_sandbox.cpp:595]: Done in #24690.

**Comment by jonatack** [comment]
Backported to v23.0 in #24512

**Pull request merged**